feat: Add `maxCount` parameter to `Badge.count` constructor.

iamtoricool opened the pull request:
feat: Add `maxCount` parameter to `Badge.count` constructor

This enhancement provides the ability to specify a **dynamic maximum limit** for the badge count, offering greater flexibility and eliminating the reliance on hardcoded values within the widget. This allows developers to, for example, display "99+" instead of "100" or "9+" instead of "10", improving UI consistency and preventing overly large numbers on badges.

---

## Description of Changes

This PR introduces a new optional parameter, `maxCount`, to the `Badge.count` constructor. This provides a flexible way to define a numerical upper bound for the badge display.

Key aspects of this change:
* **Dynamic Limits:** Developers can now set custom maximum values (e.g., 9, 99, 999) that, when exceeded, will display the `maxCount` value followed by a `+` symbol (e.g., "9+", "99+").
* **Improved UI/UX:** Prevents large, potentially overflowing numbers from appearing on badges, maintaining visual integrity and a cleaner user interface.
* **Reduced Hardcoding:** Centralizes the logic for maximum badge values within the `Badge` widget itself, reducing the need for manual checks and conditional rendering outside the widget.
* **Comprehensive Documentation:** Includes updated API documentation for the `maxCount` parameter, clarifying its usage, behavior, and the visual outcome when the maximum is reached.

This change aligns with Flutter's principles of providing highly customizable and intuitive widgets.

---

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] I followed the [breaking change policy] and added [Data Driven Fixes] where supported. (Assuming no breaking changes)
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

iamtoricool committed:
feat: Add `maxCount` parameter to `Badge.count` constructor.
- This allows for dynamic maximum limits on badge counts, removing the need for hardcoded values.

iamtoricool committed:
docs: Update `Badge.count` documentation for `maxCount`.
- Adds comprehensive API documentation for the newly introduced `maxCount` parameter.

iamtoricool committed:
Merge branch 'master' into master

QuncCccccc reviewed (commented):
Thank you for your contribution! Could you help add some unit tests to verify the code change?

iamtoricool committed:
Merge branch 'flutter:master' into master

iamtoricool committed:
Adds Badge maxCount example

Adds an example demonstrating the `maxCount` property of the `Badge` widget.

The example showcases how the badge displays a maximum count with an overflow indicator when the actual count exceeds the specified maximum.

iamtoricool committed:
Merge branch 'master' into master

iamtoricool commented:
![image](https://github.com/user-attachments/assets/58aca6b4-c7b4-47a7-912a-9df104335ee8)

> Thank you for your contribution! Could you help add some unit tests to verify the code change?

Hey, thanks for the feedback! I’ve added a `BadgeMaxCountExampleApp` demo and a widget test that covers:

* Overflow case (`1000` with `maxCount: 99` shows “99+”)
* Within‑range counts
* Default maxCount (values over 999 render as “999+”)

You’ll find them in `examples/api/lib/material/badge/badge.1.dart` and `examples/api/test/material/badge/badge.1_test.dart`. Let me know if you’d like any changes! 😊

QuncCccccc commented:
>  I’ve added a BadgeMaxCountExampleApp demo and a widget test that covers

Thanks a lot for adding them! I actually think it's not necessary to add an example just for the new parameter. What I mean in my last comment is to add unit tests in `packages/flutter/test/material/badge_test.dart`, sorry if it's misleading:) Please let me know if any help is needed on the unit tests!

QuncCccccc commented on packages/flutter/lib/src/material/badge.dart:
```diff
@@ -184,8 +186,8 @@ class Badge extends StatelessWidget {
     if (hasLabel) {
-      final double minSize =
-          effectiveWidthOffset = largeSize ?? badgeTheme.largeSize ?? defaults.largeSize!;
+      final double minSize = effectiveWidthOffset =
```
Seems this change and below are just formatting. We might need to revert these changes to fix the failed `linux_analyer` test:)

iamtoricool committed:
Fix(material): Correctly display badge count with maxCount

The `Badge.count` constructor now accurately displays the badge label based on the `count` and `maxCount` properties.

- When `count` is less than or equal to `maxCount`, the badge displays the `count` value.
- When `count` exceeds `maxCount`, the badge displays '[maxCount]+'.

This commit also includes a new test case to verify the correct behavior of `Badge.count` with `maxCount`.

The example demonstrating the `maxCount` property of the `Badge` widget has been removed.

iamtoricool committed:
Merge branch 'master' into master

iamtoricool committed:
Merge branch 'master' into master

iamtoricool committed:
Merge branch 'master' into master

pranavo72bex commented on packages/flutter/lib/src/material/badge.dart:
```diff
@@ -73,9 +75,10 @@ class Badge extends StatelessWidget {
     this.isLabelVisible = true,
     this.child,
-  }) : label = Text(count > 999 ? '999+' : '$count');
+  }) : label = Text(count > maxCount ? '$maxCount+' : '$count');
```
Move Label Creation to a Getter or Method
Right now, label assignment is inline in the initializer list. Moving this to a small private method improves readability:

```
 class Badge extends StatelessWidget {
  final int maxCount;
  late final Widget label; // ✅ Make it late so we can assign it in the constructor
```
```
}) : maxCount = displayMaxCount {
    label = _buildLabel(count, maxCount); // ✅ assign here using the method
  }
```

```
  Text _buildLabel(int count, int maxCount) {
    return Text(count > maxCount ? '$maxCount+' : '$count');
  }
```

Piinks replied:
I am not sure this is necessary, but thanks for the feedback!

pranavo72bex replied:
Just to improve code readability 😅

pranavo72bex commented on packages/flutter/test/material/badge_test.dart:
```diff
@@ -438,4 +438,34 @@ void main() {
+      );
+    }
+
+    await tester.pumpWidget(buildFrame(5, 99));
```
Add Edge Case Test for Negative Count
Even though it’s unlikely to happen in production, explicitly asserting that negative count values throw an AssertionError or are clamped would increase robustness.

pranavo72bex commented on packages/flutter/lib/src/material/badge.dart:
```diff
@@ -57,11 +57,13 @@ class Badge extends StatelessWidget {
+  /// For example, if [count] is 1000 and [maxCount] is 99, the label will display '99+'.
   Badge.count({
     super.key,
     this.backgroundColor,
```
@QuncCccccc
Would it be possible to add a semanticsLabel to the Badge widget to improve accessibility?

Currently, the badge may not be accessible to screen readers, especially when it includes dynamic content like counts or status indicators.

iamtoricool committed:
Merge branch 'master' into master

QuncCccccc reviewed (approved):
LGTM! Thank you for your contribution!

Piinks commented on packages/flutter/lib/src/material/badge.dart:
```diff
@@ -73,9 +75,10 @@ class Badge extends StatelessWidget {
     this.isLabelVisible = true,
     this.child,
-  }) : label = Text(count > 999 ? '999+' : '$count');
+  }) : label = Text(count > maxCount ? '$maxCount+' : '$count');
```
Adding the assertion here that we're non-negative is probably the right spot.

iamtoricool replied:
> Adding the assertion here that we're non-negative is probably the right spot.

Awesome, thanks for the confirmation! I've added the `assert(count >= 0, ...)` and `assert(maxCount > 0, ...)` checks in the constructor, and expanded our unit tests to include negative, zero, and invalid `maxCount` cases. The updated code and tests are pushed—please let me know if anything else needs tweaking! 😊

iamtoricool committed:
Merge branch 'flutter:master' into master

iamtoricool committed:
feat(material): add assertions and zero count display to Badge.count

This commit adds assertions to `Badge.count` to ensure that the `count` is non-negative and `maxCount` is positive. It also ensures that "0" is displayed when the count is zero.

victorsanni reviewed (approved):
LGTM. Thanks for the PR.
